### PR TITLE
[TS] LPS-175783

### PR DESCRIPTION
**Pull request by ChrisKian**
Resent from https://github.com/liferay-core-infra/liferay-portal/pull/3510

https://issues.liferay.com/browse/LPS-175783

Reworked solution based off [comment here](https://github.com/liferay-core-infra/liferay-portal/pull/3510#issuecomment-1439746509).  I've found the solution works as expected while still preserving the functionality in [LPS-44217](https://issues.liferay.com/browse/LPS-44217) and [LPS-157252](https://issues.liferay.com/browse/LPS-157252).

I only have two concerns:
1. When accessing a site virtual host with a defined locale, the site default locale is used by default, not the virtual host defined locale.  This could be intended, but I'm not sure.
2. I'm still unsure how to properly test for the scenario in LPS-175783

Any help is greatly appreciated, thanks!

CC @tinatian @yuhai

### CODE REVIEW

**Comment by yuhai** [comment]
Hi @ChrisKian,

> 1. When accessing a site virtual host with a defined locale, the site default locale is used by default, not the virtual host defined locale. This could be intended, but I'm not sure.

I think this is regression issue since we removed the [codes](https://github.com/ChrisKian/liferay-portal/blob/LPS-175783/portal-impl/src/com/liferay/portal/util/PortalInstances.java#L612-L625). We should keep the behavior consistent. When access a site virtual host before login, the default local should be site virtual host's locale. Instead of removing the [codes](https://github.com/ChrisKian/liferay-portal/blob/LPS-175783/portal-impl/src/com/liferay/portal/util/PortalInstances.java#L612-L625), adding a **WebKeys.VIRTUAL_HOST_LANGUAGE_ID** filed and change the [codes ](https://github.com/ChrisKian/liferay-portal/blob/LPS-175783/portal-impl/src/com/liferay/portal/util/PortalInstances.java#L623-L625) for `httpServletRequest.setAttribute(WebKeys.VIRTUAL_HOST_LANGUAGE_ID, languageId);`. And then add the below logic to the end of I18nFilter#getRequestedLanguageId().
```
if (Validator.isNull(requestedLanguageId)) {
			String languageId = (String)httpServletRequest.getAttribute(
				WebKeys.VIRTUAL_HOST_LANGUAGE_ID);
	
			if (languageId != null) {
				return languageId;
			}	
		}
```
In the PortalInstances class, please keep removing the [codes ](https://github.com/ChrisKian/liferay-portal/blob/LPS-175783/portal-impl/src/com/liferay/portal/util/PortalInstances.java#L588-L608).

> 2. I'm still unsure how to properly test for the scenario in LPS-175783

Please add one test(like testEnglishUserWithSpainVirtualHostAndCookieAlgorithm3) to I18nFilterTest to expose the issue. In this test, the logic just like:

a. create a VirtualHost with spain language for Liferay Site (or new Site, the default langauge is en) and set the VirtualHost to MockHttpServletRequest (Refer to PortalInstancesTest#testGetI18nLanguageId()).
**b**, httpServletRequest.setAttribute(WebKeys.USER, admin user(language is en));
c. Invoke portal.getCompanyId(MockHttpServletRequest) to set WebKeys.VIRTUAL_HOST_LANGUAGE_ID to MockHttpServletRequest.
d. Invoke I18nFilter#isFilterEnabled(MockHttpServletRequest, null).
**Assert isFilterEnabled() return true, the original issue will return false.**
e. test prependI18nLanguageId() returned value.

Finally, the pr order should be:
1. Expose the issue.
2. Fix the issue
3. Please add one additional test method "testGuestEnglishWithSpainVirtualHostAndCookieAlgorithm3" (no need to set **step b**) to cover the added fix logic in the I18nFilter class.

You may rename my mentioned test methods name if you think they are not good.

In addition, you also need to modify PortalInstancesTest#testGetI18nLanguageId().

Thanks,
Hai

**Comment by ChrisKian** [comment]
Hey @yuhai,
I've made changes based off your last comment.  The main differences are your step 1 and 3 are merged ([first commit](https://github.com/liferay-core-infra/liferay-portal/pull/3618/commits/bbf7e1f0fa3916574e8f1c41acca079708c2483d)), I've added [an optional commit which strengthens and simplifies the test suite (commit 2)](https://github.com/liferay-core-infra/liferay-portal/pull/3618/commits/227d04ed351eafcd83624170f45b321d7496c978), and [I fix the PortalInstancesTest class as the final commit](https://github.com/liferay-core-infra/liferay-portal/pull/3618/commits/cc99b376aeaf74ae72635801c1ede93536b493e5).

Please let me know if you have any questions, thanks!

**Comment by ChrisKian** [comment]
ci:test:sf

**Comment by ChrisKian** [comment]
ci:test:relevant

**Comment by Kvale1733** [comment]
@ChrisKian Regressions found
com.liferay.portal.servlet.filters.absoluteredirects.test.VirtualHostAbsoluteRedirectsFilterTest
```
com.liferay.portal.servlet.filters.absoluteredirects.test.VirtualHostAbsoluteRedirectsFilterTest > testMultipleVirtualHostsWithLocale FAILED
    java.lang.AssertionError: expected:<es_ES> but was:<null>
    	at org.junit.Assert.fail(Assert.java:89)
    	at org.junit.Assert.failNotEquals(Assert.java:835)
    	at org.junit.Assert.assertEquals(Assert.java:120)
    	at org.junit.Assert.assertEquals(Assert.java:146)
    	at com.liferay.portal.servlet.filters.absoluteredirects.test.VirtualHostAbsoluteRedirectsFilterTest.testMultipleVirtualHostsWithLocale(VirtualHostAbsoluteRedirectsFilterTest.java:106)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
    	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)

```

```
com.liferay.portal.servlet.filters.absoluteredirects.test.VirtualHostAbsoluteRedirectsFilterTest > testSingleVirtualHostWithLocale FAILED
    java.lang.AssertionError: expected:<de_DE> but was:<null>
    	at org.junit.Assert.fail(Assert.java:89)
    	at org.junit.Assert.failNotEquals(Assert.java:835)
    	at org.junit.Assert.assertEquals(Assert.java:120)
    	at org.junit.Assert.assertEquals(Assert.java:146)
    	at com.liferay.portal.servlet.filters.absoluteredirects.test.VirtualHostAbsoluteRedirectsFilterTest.testSingleVirtualHostWithLocale(VirtualHostAbsoluteRedirectsFilterTest.java:162)
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```

[Log](https://gist.github.com/Kvale1733/8407c07039df21ffccaef134c6c603c1)

**Comment by ChrisKian** [comment]
ci:test:sf

**Comment by ChrisKian** [comment]
ci:test:relevant

**Comment by ChrisKian** [comment]
ci:test:sf

**Comment by ChrisKian** [comment]
ci:test:relevant

**Comment by catarineaguiar** [comment]
no regressions found.

File: modules/apps/portal/portal-servlet-test/src/testIntegration/java/com/liferay/portal/servlet/filters/i18n/test/I18nFilterTest.java [hunk 1]
```diff
@@ -206,6 +250,16 @@ private String _getPrependI18nLanguageId(
...
+		PortalInstances.getCompanyId(_mockHttpServletRequest);
+
+		Assert.assertTrue(
+			ReflectionTestUtil.invoke(
+				_i18nFilter, "isFilterEnabled",
+				new Class<?>[] {
+					HttpServletRequest.class, HttpServletResponse.class
+				},
+				_mockHttpServletRequest, _mockHttpServletResponse));
+
```
Comment (yuhai) [inline modules/apps/portal/portal-servlet-test/src/testIntegration/java/com/liferay/portal/servlet/filters/i18n/test/I18nFilterTest.java:262]: @ChrisKian, please simplify this. You may directly invoke isFilterEnabled().

```
Assert.assertTrue(
	_i18nFilter.isFilterEnabled(
		_mockHttpServletRequest, _mockHttpServletResponse));
```

File: modules/apps/portal/portal-servlet-test/src/testIntegration/java/com/liferay/portal/servlet/filters/i18n/test/I18nFilterTest.java [hunk 2]
```diff
@@ -206,6 +250,16 @@ private String _getPrependI18nLanguageId(
 				_mockHttpServletResponse.getCookies());
 		}
 
+		PortalInstances.getCompanyId(_mockHttpServletRequest);
+
```
Comment (yuhai) [inline modules/apps/portal/portal-servlet-test/src/testIntegration/java/com/liferay/portal/servlet/filters/i18n/test/I18nFilterTest.java:254]: @ChrisKian, please remove `PortalInstances.getCompanyId(_mockHttpServletRequest);`.  Since VirtualHost related tests's cookieLocale is not null and _language.updateCookie() will invoke `PortalInstances.getCompanyId(_mockHttpServletRequest)`

File: modules/apps/portal/portal-servlet-test/src/testIntegration/java/com/liferay/portal/servlet/filters/i18n/test/I18nFilterTest.java [hunk 3]
```diff
@@ -173,17 +190,44 @@ public void testGuestSpanishSessionWithSpanishCookieAlgorithm3()
...
+				RandomTestUtil.randomString(6) + "." +
+					RandomTestUtil.randomString(3);
+
+			LayoutSet layoutSet = _group.getPublicLayoutSet();
+
+			_virtualHostLocalService.updateVirtualHosts(
+				_group.getCompanyId(), layoutSet.getLayoutSetId(),
+				TreeMapBuilder.put(
+					StringUtil.toLowerCase(layoutHostname), "es_ES"
+				).build());
```
Comment (yuhai) [inline modules/apps/portal/portal-servlet-test/src/testIntegration/java/com/liferay/portal/servlet/filters/i18n/test/I18nFilterTest.java:220]: @ChrisKian, please replace "es_ES" with `LocaleUtil.toLanguageId(virtualHostLocale)`

**Comment by yuhai** [comment]
@ChrisKian, for the I18nFilterTest, it doesn't cover I18nFilter's fix since sessionLocale always is not Null. I think you may add the test cases "sessionLocale is null" to cover the [fix ](https://github.com/liferay-core-infra/liferay-portal/blob/7a2704cc7da11cf727ff92e751eb3e0d249f088b/portal-impl/src/com/liferay/portal/servlet/filters/i18n/I18nFilter.java#L253-L256). 

Please also check my above comments and then resend it.

Thanks,
Hai

**Comment by ChrisKian** [comment]
Resent here: https://github.com/liferay-core-infra/liferay-portal/pull/3699
